impl/client/src/main/javascript/web/dojo/pentaho/common/PageControl.js

```diff
@@ -30,161 +30,185 @@
+            },
+
+            setPageCount: function (pageCount) {
+                if (typeof (pageCount) == 'number') {
+                    this.pageTotal = pageCount;
+                } else {
+                    this.pageTotal = 1; // Reset to default
+                }
+                this.pageTotalSpan.innerHTML = '/ ' + pageCount;
+                this._updatePageButtonState();
+            },
+
+            reset: function () {
+                this._updateInternalPageNumber(1);
+            },
+
+            _localize: function () {
+                this.prevPageControl.setAttribute("title", this.getLocaleString("PageControlPreviousPage_title"));
+                this.nextPageControl.setAttribute("title", this.getLocaleString("PageControlNextPage_title"));
+            },
+
+            _updatePageButtonState: function () {
+                this.pagePrevDisabled = this.pageNumber === 1;
+                this.pageNextDisabled = this.pageTotal === this.pageNumber;
+                if (this.pagePrevDisabled) {
+                    domClass.add(this.prevPageControl, "pc_pagePrevDisabled");
+                    this.prevPageControl.setAttribute("tabindex", "-1");
+                } else {
+                    domClass.remove(this.prevPageControl, "pc_pagePrevDisabled");
+                    this.prevPageControl.setAttribute("tabindex", "0");
+                }
+                if (this.pageNextDisabled) {
+                    domClass.add(this.nextPageControl, "pc_pageNextDisabled");
+                    this.nextPageControl.setAttribute("tabindex", "-1");
+                } else {
+                    domClass.remove(this.nextPageControl, "pc_pageNextDisabled");
+                    this.nextPageControl.setAttribute("tabindex", "0");
+                }
+            },
+
+            _changePageNumber: function () {
+                if (this.ignorePageNumberUpdate) {
+                    return;
+                }
+                if (!this.pageNumberInput.isValid()) {
+                    this._resetPageNumber();
+                    return;
+                }
+                this.setPageNumber(this.pageNumberInput.get("value"));
+            },
+
+            _updateInternalPageNumber: function (pageNumber) {
+                if (this.pageNumber === pageNumber) {
+                    return;
+                }
+
+                this.ignorePageNumberUpdate = true;
+                try {
+                    if (this.changePageNumberCallback) {
+                        this.changePageNumberCallback(pageNumber);
+                    }
+                    this.pageNumber = pageNumber;
+                    this.pageNumberInput.set("value", this.pageNumber);
+                    this._updatePageButtonState();
+                } finally {
+                    this.ignorePageNumberUpdate = false;
+                }
+                this.pageNumber = pageNumber;
+                this.pageNumberInput.set("value", this.pageNumber);
```

Comment:
Why not define a protected "_setPageNumberInput" method which manages the reentry control, instead of doing it everywhere the pageNumberInput is explicitly set from code?

Also, this way, the flag is set to a true value before actually needing to, making its "raison d'étre" (reason to exist) less clear.

I also find it strange that the flag is only set on the first time that the pageNumberInput is set, and not on the second time (this line). I believe it should be managed in both cases, to allow `_changePageNumber` to react only to changes initiated by the user by editing the text box directly, and to allow ignoring other code initiated changes.

Reply:
I now realize all these comments apply equally to the https://github.com/pentaho/pentaho-platform-plugin-interactive-reporting/blob/wcag/web/src/main/javascript/pir/pir/PageControl.js you based these changes on...